src/coreclr/minipal/Unix/doublemapping.cpp

```diff
@@ -47,6 +47,90 @@ static const off_t MaxDoubleMappedSize = UINT_MAX;
 
 #endif // TARGET_OSX
 
+#if defined(TARGET_AMD64) && !defined(TARGET_OSX)
+
+extern "C" int VerifyDoubleMapping1();
+extern "C" void VerifyDoubleMapping1_End();
+extern "C" int VerifyDoubleMapping2();
+extern "C" void VerifyDoubleMapping2_End();
+
+// Verify that the double mapping works correctly, including cases when the executable code page is modified after
+// the code is executed. 
+bool VerifyDoubleMapping(int fd)
+{
+    bool result = false;
+    void *mapperHandle = (void*)(size_t)fd;
+    void *pCommittedPage = NULL;
+    void *pWriteablePage = NULL;
+    int testCallResult;
+
+    typedef int (*VerificationFunctionPtr)();
+    VerificationFunctionPtr pVerificationFunction;
+
+    size_t pageSize = getpagesize();
+
+    void *pExecutablePage = VMToOSInterface::ReserveDoubleMappedMemory(mapperHandle, 0, pageSize, NULL, NULL);
+    
+    if (pExecutablePage == NULL)
+    {
+        goto Cleanup;
+    }
+
+    pCommittedPage = VMToOSInterface::CommitDoubleMappedMemory(pExecutablePage, pageSize, true);
+    if (pCommittedPage == NULL)
+    {
+        goto Cleanup;
+    }
+
+    pWriteablePage = VMToOSInterface::GetRWMapping(mapperHandle, pCommittedPage, 0, pageSize);
+    if (pWriteablePage == NULL)
+    {
+        goto Cleanup;
+    }
+
+    // First copy a method of a simple function that returns 1 into the writeable mapping
+    memcpy(pWriteablePage, (void*)VerifyDoubleMapping1, (char*)VerifyDoubleMapping1_End - (char*)VerifyDoubleMapping1);
+    pVerificationFunction = (VerificationFunctionPtr)pExecutablePage;
+    // Invoke the function via the executable mapping. It should return 1.
+    testCallResult = pVerificationFunction();
+    if (testCallResult != 1)
+    {
+        goto Cleanup;
+    }
+
+    VMToOSInterface::ReleaseRWMapping(pWriteablePage, pageSize);
+    pWriteablePage = VMToOSInterface::GetRWMapping(mapperHandle, pCommittedPage, 0, pageSize);
+    if (pWriteablePage == NULL)
+    {
+        goto Cleanup;
+    }
+
+    // Now overwrite the first function by a second one that returns 2 using the writeable mapping
+    memcpy(pWriteablePage, (void*)VerifyDoubleMapping2, (char*)VerifyDoubleMapping2_End - (char*)VerifyDoubleMapping2);
+    pVerificationFunction = (VerificationFunctionPtr)pExecutablePage;
+    testCallResult = pVerificationFunction();
+    // Invoke the function via the executable mapping again. It should return 2 now.
+    // This doesn't work when running x64 code in docker on macOS Arm64 where the code is not re-translated by Rosetta
```

Comment:
I am wondering whether it is possible to make it work by issuing flush instruction cache syscall. Is there a syscall like on Linux x64?

Similar to what we do for Windows: https://github.com/dotnet/runtime/blob/362a95d53d7266ee9b7452e351d07845dcc1edd2/src/coreclr/vm/amd64/cgencpu.h#L588-L605

Reply:
Apple folks told me in the past that Rosetta doesn't support this scenario, but I can add the cache flushing. In fact, it should be in this testing method anyways.

Reply:
To be specific, the problem is that Rosetta JITs the x64 code into arm64 on first execution and uses the result ever after. Since it doesn't get any notification on the executable code modification via a secondary mapping, it doesn't re-JIT it and keeps using the old code.

Reply:
As for the cache flushing, there is the `__builtin___clear_cache`, but it doesn't generate any code on x64. And there is no syscall to do that on Linux.